### ISSUE TITLE
Files are not being deleted, also when "dir"-field is set

### ISSUE THREAD

**Issue by TheMiller2015** [josegonzalez/cakephp-upload#406]
Hello,

I use the upload-plugin as a polymorphic upload system. I have a database-structure like this: "table: attachments; fields: id|section|filesize|filepath|filename".

The configuration of the upload-behaviour is like this (the input-field is called "filename"):

```
$this->addBehavior('Josegonzalez/Upload.Upload', [
            'filename' => [
                'path' => 'webroot{DS}files{DS}{year}{DS}{month}{DS}{day}',
                'fields' => [
                    'dir' => 'filepath',
                    'size' => 'filesize',
                    'type' => 'datatype',
                ],
                'pathProcessor' => '\App\Controller\Lib\Upload\Paths\DefaultProcessor',
                'keepFilesOnDelete' => false,         
            ],            
        ]);
```

The files are being stored, viewable and downloadable. But when I delete an attachment via attachment-controller, the file remains on the hdd. When I delete an other entity (like user), neither the file wille be deleted nor the attachment-record. 

When I use a event-function in the attachments-table (like below), it's being deleted.

```
/**
     * Delete the files from storage and differ deletion by "section" of
     * entity (e.g. pictures have many files because of thumbnails etc.) 
     *
     */
    public function beforeDelete($event, $entity, $options) {
        if(!unlink("..".DS.$entity->filepath.DS.$entity->filename)) {
            $event->stopPropagation();
        }
    }
```

I think, it is because of "webroot/". When I want to download a file, I have to strip off the webroot-string with substring like this:

```
public function view($id = null)
    {
        $datafile = $this->Datafiles->get($id);
        $this->response->file(**substr($datafile->filepath, 8, 100)**.DS.$datafile->filename, [
            'name' => $datafile->filename_orig,
            'download' => true,
        ]);
        return $this->response;
    }
```

Or could the association be the problem? But saving the attachment works great without extra-coding.

Did I configure something wrong? Thanks in advance!

**Comment by ADmad** [comment]
> When I delete an other entity (like user), neither the file wille be deleted nor the attachment-record.

Make sure you have "dependent" and "cascadeCallbacks" options are set to true in the association.

**Comment by TheMiller2015** [comment]
Ahhh, this could do the trick. I'll test it tomorrow morning and post my results. 

Many thanks for now ;)

**Comment by josegonzalez** [comment]
File deletion definitely works, though I'm open to seeing a pull request with a test proving otherwise (I can fix once I have a testcase).

**Comment by TheMiller2015** [comment]
I don't know how to write tests in cake yet. My association is fixed now and the related attachment entity in database is being deleted, but not the file. I do have to delete it manually in beforeDelete while manipulating the filepath (see below). I'm trying to explain, what I did. Shall I post a gist?

I have a relationship like this: ChatMessage hasOne Attachment | One Attachment belongsTo one ChatMessage.

In ChatMessage-Table (mySQL) I have a field "attachment_id". Is is not being filled while adding a ChatMessage (btw: Is this correct? I think it isn't needed).

In the Attachment-Table, cake stores the path like this: webroot{DS}files{DS}{year}{DS}{month}{DS}{day}, where "webroot/" is written to the entity in db. So cake can't find/delete the file. Therefore, I removed the file manually in "beforeDelete", stripped off the "webroot/"-part of the string and so came up with this:  `if(!unlink("..".DS.$entity->filepath.DS.$entity->filename)) {`.

But when I don't save the "webroot/"-path, the file will not be saved to the correct directory after uploading. 

I am sure, that the save- and delete-methods are working, but I don't know what's the catch here. I read the docs slowly, I searched the issues, but I don't know what's wrong.

Thanks again.

**Comment by josegonzalez** [comment]
[Here](https://github.com/josegonzalez/cakephp-upload/blob/master/tests/TestCase/Model/Behavior/UploadBehaviorTest.php#L330) is an example testcase for file deletion that you might be able to copy/modify to replicate your issue.

**Comment by josegonzalez** [comment]
Alternatively, please package up a sample application (with a schema of course) that replicates your issue, otherwise i'll have a hard time testing it.

**Comment by TheMiller2015** [comment]
Ok thanks. I just started to create a package and I saw, that I did a "mistake". At first, I forgot the trailing "{DS}" in the path-configuration of your upload-behavior. (I corrected this).

Then I realised, that my database-field is called "filename". I wanted to aliase the html-field so that it is not called "filename" (for "security" reasons). Therefore, I called this "attachment" and also the upload-behavior I set to `attachment => [....`. Saving the data works, but while deleting a record, it seems not to find the correct field (of course...).

When I name the configuration-field "filename", the html-form-field "filename", everything works just fine - even with a custom pathProcessor. So I guess the alias was my big big mistake. 

Generally - may it be a security issue, when the html-form-fields are named similar to the database-fields as an attacker could "guess" the db-schema? I think, this naming-convention is in cake the default one. Is it safe, or should I alias the fields?

Thanks for your great help! (And also the awesome work on cake and this plugin).
